jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1127 @@
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public E removeEdge(V sourceVertex, V targetVertex)
+    {
+        synchronized (mutex) {
+            E e = cacheStrategy.removeEdge(sourceVertex, targetVertex);
+            if (e != null)
+                edgeSetModified();
+            return e;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean removeVertex(V v)
+    {
+        synchronized (mutex) {
+            if (cacheStrategy.removeVertex(v)) {
+                edgeSetModified();
+                vertexSetModified();
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString()
+    {
+        synchronized (mutex) {
+            return super.toString();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<V> vertexSet()
+    {
+        synchronized (mutex) {
+            return allVerticesSet;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public V getEdgeSource(E e)
+    {
+        synchronized (mutex) {
+            return super.getEdgeSource(e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public V getEdgeTarget(E e)
+    {
+        synchronized (mutex) {
+            return super.getEdgeTarget(e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public double getEdgeWeight(E e)
+    {
+        synchronized (mutex) {
+            return super.getEdgeWeight(e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setEdgeWeight(E e, double weight)
+    {
+        synchronized (mutex) {
+            super.setEdgeWeight(e, weight);
+        }
+    }
+
+    /**
+     * Return whether the graph uses cache for <code>edgesOf</code>, <code>incomingEdgesOf</code> and
+     * <code>outgoingEdgesOf</code> methods.
+     * @return <tt>true</tt> if cache is in use, <tt>false</tt> if cache is not in use.
+     */
+    public boolean isCacheEnabled()
+    {
+        synchronized (mutex) {
+            return cacheStrategy.isCacheEnabled();
+        }
+    }
+
+    /**
+     * Set the cache strategy for <code>edgesOf</code>, <code>incomingEdgesOf</code> and
+     * <code>outgoingEdgesOf</code> methods.
+     * @param cacheEnabled a flag whether to use cache for those methods, if <tt>true</tt>, cache
+     *        will be used for those methods, otherwise cache will not be used.
+     */
+    public void setCache(boolean cacheEnabled)
+    {
+        synchronized (mutex) {
+            if (cacheEnabled == isCacheEnabled())
+                return;
+            if (cacheEnabled)
+                cacheStrategy = new CacheAccess();
+            else
+                cacheStrategy = new NoCache();
+        }
+    }
```

Comment:
Do we really need this method when we already have a constructor that accepts a `cacheEnabled` parameter? I can't imagine that having _more_ mutable state in this class is going to help things from a user's perspective. :/

Reply:
Sorry, by "more mutable state", I meant having a method that would act as another venue to mutate some state inside this class.

Reply:
If in a period of time, the user is more likely to modify the graph(when initially building the graph, add operations will be more offen than get operations.), the NoCache strategy will be better, if the user is more likely to visit the graph and have low memory requirements, CacheAccess will be better. With this in mind, I did this design.

Reply:
Ah okay, I hadn't thought of those scenarios, so this design seems more reasonable to me now.

How about having this method return `this` so as to allow users to use it in a more fluent or builder-like manner?

Reply:
I get that. If do that, then if user has a large of edges or vertices to remove, user can can only do like this:
```
graph.setCache(false).removeAllEdges(collection);
```

Reply:
Sounds good to me. :)